// telegram/tl/core/GZIPPacked.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GZIPPacked = void 0;
const __1 = require("../");
const pako_1 = require("pako");
class GZIPPacked {
    constructor(data) {
        this.data = data;
        this.CONSTRUCTOR_ID = 0x3072cfa1;
        this.classType = 'constructor';
    }
    static gzipIfSmaller(contentRelated, data) {
        return __awaiter(this, void 0, void 0, function* () {
            if (contentRelated && data.length > 512) {
                const gzipped = yield (new GZIPPacked(data)).toBytes();
                if (gzipped.length < data.length) {
                    return gzipped;
                }
            }
            return data;
        });
    }
    static gzip(input) {
        return Buffer.from(input);
        // TODO this usually makes it faster for large requests
        //return Buffer.from(deflate(input, { level: 9, gzip: true }))
    }
    static ungzip(input) {
        return Buffer.from(pako_1.inflate(input));
    }
    toBytes() {
        return __awaiter(this, void 0, void 0, function* () {
            const g = Buffer.alloc(4);
            g.writeUInt32LE(GZIPPacked.CONSTRUCTOR_ID, 0);
            return Buffer.concat([
                g,
                __1.serializeBytes(yield GZIPPacked.gzip(this.data)),
            ]);
        });
    }
    static read(reader) {
        return __awaiter(this, void 0, void 0, function* () {
            const constructor = reader.readInt(false);
            if (constructor !== GZIPPacked.CONSTRUCTOR_ID) {
                throw new Error('not equal');
            }
            return yield GZIPPacked.gzip(reader.tgReadBytes());
        });
    }
    static fromReader(reader) {
        return __awaiter(this, void 0, void 0, function* () {
            const data = reader.tgReadBytes();
            return new GZIPPacked(yield GZIPPacked.ungzip(data));
        });
    }
}
exports.GZIPPacked = GZIPPacked;
GZIPPacked.CONSTRUCTOR_ID = 0x3072cfa1;
GZIPPacked.classType = 'constructor';
