// telegram/tl/core/RPCResult.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RPCResult = void 0;
const api_1 = require("../api");
const _1 = require("./");
class RPCResult {
    constructor(reqMsgId, body, error) {
        this.CONSTRUCTOR_ID = 0xf35c6d01;
        this.reqMsgId = reqMsgId;
        this.body = body;
        this.error = error;
        this.classType = 'constructor';
    }
    static fromReader(reader) {
        return __awaiter(this, void 0, void 0, function* () {
            const msgId = reader.readLong();
            const innerCode = reader.readInt(false);
            if (innerCode === api_1.Api.RpcError.CONSTRUCTOR_ID) {
                return new RPCResult(msgId, undefined, api_1.Api.RpcError.fromReader(reader));
            }
            if (innerCode === _1.GZIPPacked.CONSTRUCTOR_ID) {
                return new RPCResult(msgId, (yield _1.GZIPPacked.fromReader(reader)).data);
            }
            reader.seek(-4);
            // This reader.read() will read more than necessary, but it's okay.
            // We could make use of MessageContainer's length here, but since
            // it's not necessary we don't need to care about it.
            return new RPCResult(msgId, reader.read(), undefined);
        });
    }
}
exports.RPCResult = RPCResult;
RPCResult.CONSTRUCTOR_ID = 0xf35c6d01;
RPCResult.classType = 'constructor';
