// telegram/extensions/PromisedNetSockets.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromisedNetSockets = void 0;
const net_1 = require("net");
const async_mutex_1 = require("async-mutex");
const mutex = new async_mutex_1.Mutex();
const closeError = new Error('NetSocket was closed');
class PromisedNetSockets {
    constructor() {
        this.client = undefined;
        this.closed = true;
        this.stream = Buffer.alloc(0);
    }
    readExactly(number) {
        return __awaiter(this, void 0, void 0, function* () {
            let readData = Buffer.alloc(0);
            while (true) {
                const thisTime = yield this.read(number);
                readData = Buffer.concat([readData, thisTime]);
                number = number - thisTime.length;
                if (!number) {
                    return readData;
                }
            }
        });
    }
    read(number) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.closed) {
                throw closeError;
            }
            yield this.canRead;
            if (this.closed) {
                throw closeError;
            }
            const toReturn = this.stream.slice(0, number);
            this.stream = this.stream.slice(number);
            if (this.stream.length === 0) {
                this.canRead = new Promise(resolve => {
                    this.resolveRead = resolve;
                });
            }
            return toReturn;
        });
    }
    readAll() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.closed || !(yield this.canRead)) {
                throw closeError;
            }
            const toReturn = this.stream;
            this.stream = Buffer.alloc(0);
            this.canRead = new Promise(resolve => {
                this.resolveRead = resolve;
            });
            return toReturn;
        });
    }
    /**
     * Creates a new connection
     * @param port
     * @param ip
     * @returns {Promise<void>}
     */
    connect(port, ip) {
        return __awaiter(this, void 0, void 0, function* () {
            this.stream = Buffer.alloc(0);
            this.client = new net_1.Socket();
            this.canRead = new Promise(resolve => {
                this.resolveRead = resolve;
            });
            this.closed = false;
            return new Promise((resolve, reject) => {
                if (this.client) {
                    this.client.connect(port, ip, () => {
                        this.receive();
                        resolve(this);
                    });
                    this.client.on('error', reject);
                    this.client.on('close', () => {
                        if (this.client && this.client.destroyed) {
                            if (this.resolveRead) {
                                this.resolveRead(false);
                            }
                            this.closed = true;
                        }
                    });
                }
            });
        });
    }
    write(data) {
        if (this.closed) {
            throw closeError;
        }
        if (this.client) {
            this.client.write(data);
        }
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.client) {
                yield this.client.destroy();
                this.client.unref();
            }
            this.closed = true;
        });
    }
    receive() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.client) {
                this.client.on('data', (message) => __awaiter(this, void 0, void 0, function* () {
                    const release = yield mutex.acquire();
                    try {
                        let data;
                        //CONTEST BROWSER
                        this.stream = Buffer.concat([this.stream, message]);
                        if (this.resolveRead) {
                            this.resolveRead(true);
                        }
                    }
                    finally {
                        release();
                    }
                }));
            }
        });
    }
}
exports.PromisedNetSockets = PromisedNetSockets;
