// telegram/network/connection/Connection.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObfuscatedConnection = exports.PacketCodec = exports.Connection = void 0;
const extensions_1 = require("../../extensions");
const extensions_2 = require("../../extensions");
const Helpers_1 = require("../../Helpers");
/**
 * The `Connection` class is a wrapper around ``asyncio.open_connection``.
 *
 * Subclasses will implement different transport modes as atomic operations,
 * which this class eases doing since the exposed interface simply puts and
 * gets complete data payloads to and from queues.
 *
 * The only error that will raise from send and receive methods is
 * ``ConnectionError``, which will raise when attempting to send if
 * the client is disconnected (includes remote disconnections).
 */
class Connection {
    constructor(ip, port, dcId, loggers) {
        this._ip = ip;
        this._port = port;
        this._dcId = dcId;
        this._log = loggers;
        this._connected = false;
        this._sendTask = undefined;
        this._recvTask = undefined;
        this._codec = undefined;
        this._obfuscation = undefined; // TcpObfuscated and MTProxy
        this._sendArray = new extensions_2.AsyncQueue();
        this._recvArray = new extensions_2.AsyncQueue();
        this.socket = Helpers_1.IS_NODE ? new extensions_1.PromisedNetSockets() : new extensions_1.PromisedWebSockets();
        //this.socket = new PromisedWebSockets()
    }
    _connect() {
        return __awaiter(this, void 0, void 0, function* () {
            this._log.debug('Connecting');
            this._codec = new this.PacketCodecClass(this);
            yield this.socket.connect(this._port, this._ip);
            this._log.debug('Finished connecting');
            // await this.socket.connect({host: this._ip, port: this._port});
            yield this._initConn();
        });
    }
    connect() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._connect();
            this._connected = true;
            this._sendTask = this._sendLoop();
            this._recvTask = this._recvLoop();
        });
    }
    disconnect() {
        return __awaiter(this, void 0, void 0, function* () {
            this._connected = false;
            yield this._recvArray.push(undefined);
            yield this.socket.close();
        });
    }
    send(data) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._connected) {
                // this will stop the current loop
                // @ts-ignore
                yield this._sendArray(undefined);
                throw new Error('Not connected');
            }
            yield this._sendArray.push(data);
        });
    }
    recv() {
        return __awaiter(this, void 0, void 0, function* () {
            while (this._connected) {
                const result = yield this._recvArray.pop();
                // undefined = sentinel value = keep trying
                if (result && result.length) {
                    return result;
                }
            }
            throw new Error('Not connected');
        });
    }
    _sendLoop() {
        return __awaiter(this, void 0, void 0, function* () {
            // TODO handle errors
            try {
                while (this._connected) {
                    const data = yield this._sendArray.pop();
                    if (!data) {
                        this._sendTask = undefined;
                        return;
                    }
                    yield this._send(data);
                }
            }
            catch (e) {
                this._log.info('The server closed the connection while sending');
            }
        });
    }
    _recvLoop() {
        return __awaiter(this, void 0, void 0, function* () {
            let data;
            while (this._connected) {
                try {
                    data = yield this._recv();
                    if (!data) {
                        throw new Error('no data received');
                    }
                }
                catch (e) {
                    this._log.info('connection closed');
                    //await this._recvArray.push()
                    console.log(e);
                    this.disconnect();
                    return;
                }
                yield this._recvArray.push(data);
            }
        });
    }
    _initConn() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this._codec.tag) {
                yield this.socket.write(this._codec.tag);
            }
        });
    }
    _send(data) {
        return __awaiter(this, void 0, void 0, function* () {
            const encodedPacket = this._codec.encodePacket(data);
            this.socket.write(encodedPacket);
        });
    }
    _recv() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._codec.readPacket(this.socket);
        });
    }
    toString() {
        return `${this._ip}:${this._port}/${this.constructor.name.replace('Connection', '')}`;
    }
}
exports.Connection = Connection;
class ObfuscatedConnection extends Connection {
    constructor() {
        super(...arguments);
        this.ObfuscatedIO = undefined;
    }
    _initConn() {
        return __awaiter(this, void 0, void 0, function* () {
            this._obfuscation = new this.ObfuscatedIO(this);
            this.socket.write(this._obfuscation.header);
        });
    }
    _send(data) {
        return __awaiter(this, void 0, void 0, function* () {
            this._obfuscation.write(this._codec.encodePacket(data));
        });
    }
    _recv() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this._codec.readPacket(this._obfuscation);
        });
    }
}
exports.ObfuscatedConnection = ObfuscatedConnection;
class PacketCodec {
    constructor(connection) {
        this._conn = connection;
    }
    encodePacket(data) {
        throw new Error('Not Implemented');
        // Override
    }
    readPacket(reader) {
        return __awaiter(this, void 0, void 0, function* () {
            // override
            throw new Error('Not Implemented');
        });
    }
}
exports.PacketCodec = PacketCodec;
