// telegram/tl/custom/inlineResult.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InlineResult = void 0;
const api_1 = require("../api");
const __1 = require("../../");
class InlineResult {
    constructor(client, original, queryId, entity) {
        this.ARTICLE = 'article';
        this.PHOTO = 'photo';
        this.GIF = 'gif';
        this.VIDEO = 'video';
        this.VIDEO_GIF = 'mpeg4_gif';
        this.AUDIO = 'audio';
        this.DOCUMENT = 'document';
        this.LOCATION = 'location';
        this.VENUE = 'venue';
        this.CONTACT = 'contact';
        this.GAME = 'game';
        this._client = client;
        this.result = original;
        this._queryId = queryId;
        this._entity = entity;
    }
    get type() {
        return this.result.type;
    }
    get message() {
        return this.result.sendMessage;
    }
    get description() {
        return this.result.description;
    }
    get url() {
        if (this.result instanceof api_1.Api.BotInlineResult) {
            return this.result.url;
        }
    }
    get photo() {
        if (this.result instanceof api_1.Api.BotInlineResult) {
            return this.result.thumb;
        }
        else {
            return this.result.photo;
        }
    }
    get document() {
        if (this.result instanceof api_1.Api.BotInlineResult) {
            return this.result.content;
        }
        else {
            return this.result.document;
        }
    }
    click(entity, replyTo, silent = false, clearDraft = false, hideVia = false) {
        return __awaiter(this, void 0, void 0, function* () {
            if (entity) {
                entity = yield this._client.getInputEntity(entity);
            }
            else if (this._entity) {
                entity = this._entity;
            }
            else {
                throw new Error("You must provide the entity where the result should be sent to");
            }
            const replyId = replyTo ? __1.utils.getMessageId(replyTo) : undefined;
            const request = new api_1.Api.messages.SendInlineBotResult({
                peer: entity,
                queryId: this._queryId,
                id: this.result.id,
                silent: silent,
                clearDraft: clearDraft,
                hideVia: hideVia,
                replyToMsgId: replyId,
            });
            return yield this._client.invoke(request);
        });
    }
}
exports.InlineResult = InlineResult;
