// telegram/client/uploads.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendFile = exports.uploadFile = exports.CustomFile = void 0;
const tl_1 = require("../tl");
const Helpers_1 = require("../Helpers");
const Utils_1 = require("../Utils");
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
class CustomFile {
    constructor(name, size, path) {
        this.name = name;
        this.size = size;
        this.path = path;
    }
}
exports.CustomFile = CustomFile;
const KB_TO_BYTES = 1024;
const LARGE_FILE_THRESHOLD = 10 * 1024 * 1024;
const UPLOAD_TIMEOUT = 15 * 1000;
function uploadFile(client, fileParams) {
    return __awaiter(this, void 0, void 0, function* () {
        const { file, onProgress } = fileParams;
        let { workers } = fileParams;
        const { name, size } = file;
        const fileId = Helpers_1.readBigIntFromBuffer(Helpers_1.generateRandomBytes(8), true, true);
        const isLarge = size > LARGE_FILE_THRESHOLD;
        const partSize = Utils_1.getAppropriatedPartSize(size) * KB_TO_BYTES;
        const partCount = Math.floor((size + partSize - 1) / partSize);
        const buffer = Buffer.from(yield fileToBuffer(file));
        // We always upload from the DC we are in.
        const sender = yield client._borrowExportedSender(client.session.dcId);
        if (!workers || !size) {
            workers = 1;
        }
        if (workers >= partCount) {
            workers = partCount;
        }
        let progress = 0;
        if (onProgress) {
            onProgress(progress);
        }
        for (let i = 0; i < partCount; i += workers) {
            let sendingParts = [];
            let end = i + workers;
            if (end > partCount) {
                end = partCount;
            }
            for (let j = i; j < end; j++) {
                const bytes = buffer.slice(j * partSize, (j + 1) * partSize);
                sendingParts.push((() => __awaiter(this, void 0, void 0, function* () {
                    yield sender.send(isLarge
                        ? new tl_1.Api.upload.SaveBigFilePart({
                            fileId,
                            filePart: j,
                            fileTotalParts: partCount,
                            bytes,
                        })
                        : new tl_1.Api.upload.SaveFilePart({
                            fileId,
                            filePart: j,
                            bytes,
                        }));
                    if (onProgress) {
                        if (onProgress.isCanceled) {
                            throw new Error('USER_CANCELED');
                        }
                        progress += (1 / partCount);
                        onProgress(progress);
                    }
                }))());
            }
            try {
                yield Promise.race([
                    yield Promise.all(sendingParts),
                    Helpers_1.sleep(UPLOAD_TIMEOUT * workers).then(() => Promise.reject(new Error('TIMEOUT'))),
                ]);
            }
            catch (err) {
                if (err.message === 'TIMEOUT') {
                    console.warn('Upload timeout. Retrying...');
                    i -= workers;
                    continue;
                }
                throw err;
            }
        }
        return isLarge
            ? new tl_1.Api.InputFileBig({
                id: fileId,
                parts: partCount,
                name,
            })
            : new tl_1.Api.InputFile({
                id: fileId,
                parts: partCount,
                name,
                md5Checksum: '', // This is not a "flag", so not sure if we can make it optional.
            });
    });
}
exports.uploadFile = uploadFile;
function sendFile(client, entity, { file, caption, forceDocument, fileSize, progressCallback, replyTo, attributes, thumb, voiceNote, videoNote, supportStreaming }) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!file) {
            throw new Error("You need to specify a file");
        }
        if (!caption) {
            caption = "";
        }
        if (typeof file == "string") {
            file = new CustomFile(path_1.default.basename(file), fs_1.default.statSync(file).size, file);
        }
        const media = yield client.uploadFile({
            file: file,
            workers: 1,
            onProgress: progressCallback,
        });
        if (!attributes) {
            attributes = [];
        }
        let mimeType = "application/octet-stream";
        if (file instanceof CustomFile) {
            const result = (Utils_1.getAttributes(file, {
                attributes: attributes,
                forceDocument: forceDocument,
                voiceNote: voiceNote,
                videoNote: videoNote,
                supportsStreaming: supportStreaming,
                thumb: thumb
            }));
            mimeType = result.mimeType;
            attributes.push(...result.attrs);
        }
        let toSend;
        if (mimeType.startsWith("photo/")) {
            toSend = new tl_1.Api.InputMediaUploadedPhoto({
                file: media,
            });
        }
        else {
            toSend = new tl_1.Api.InputMediaUploadedDocument({
                file: media,
                mimeType: mimeType,
                attributes: attributes,
                forceFile: forceDocument,
            });
        }
        const result = yield client.invoke(new tl_1.Api.messages.SendMedia({
            peer: entity,
            media: toSend,
            replyToMsgId: replyTo,
            message: caption,
        }));
        // TODO get result
        return result;
    });
}
exports.sendFile = sendFile;
function fileToBuffer(file) {
    if (typeof File !== 'undefined' && file instanceof File) {
        return new Response(file).arrayBuffer();
    }
    else if (file instanceof CustomFile) {
        return fs_1.default.readFileSync(file.path);
    }
    else {
        throw new Error("Could not create buffer from file " + file);
    }
}
